File: 106.cpp
#include <vector>
#include <iostream>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
  
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (inorder.empty() || postorder.empty()) return nullptr;
        else return constructBinTree(inorder, 0, inorder.size()-1, postorder, 0, postorder.size()-1);
    }

private:
    TreeNode* constructBinTree(vector<int> &inorder, int startInorder,
        int endInorder, vector<int> &postorder, int startPostorder, int endPostorder) {
            cout << "endPostorder: " << endPostorder << " ";
            if (startInorder > endInorder || startPostorder > endPostorder) return nullptr;
            int rootValue = postorder[endPostorder];
            cout << rootValue;
            TreeNode *root = new TreeNode(rootValue);
            // if (startInorder == endInorder) {
            //     if (startPostorder == endPostorder && inorder[startInorder] == postorder[startPostorder]) {
            //         cout << "equal\n";
            //         return root;
            //     }
            // }
            
            int rootInorder = startInorder;
            while (rootInorder <= endInorder && inorder[rootInorder] != rootValue)
            {
                ++rootInorder;
            }
            
            int num = rootInorder - startInorder;
            cout << " " << rootInorder << " " << num << endl;

            root->left = constructBinTree(inorder, startInorder, rootInorder - 1, postorder, startPostorder, startPostorder + num -1);
            root->right = constructBinTree(inorder, rootInorder + 1, endInorder, postorder, startPostorder + num, endPostorder-1);
            return root;
        }
};

int main(int argc, char const *argv[])
{
    vector<int> inorder({1, 2});
    vector<int> postorder({2, 1});
    Solution s = Solution();
    s.buildTree(inorder, postorder);
    // cout << s.buildTree(inorder, postorder)->val << " " << 1;
    return 0;
}

File: 106.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if not inorder:
            return None
        def helper(start, end):
            if start > end:
                return None
            val = postorder.pop()
            root = TreeNode(val)
            index = inorder.index(val)
            root.right = helper(index+1, end)
            root.left = helper(start, index-1)
            return root

        return helper(0, len(inorder) - 1)
    
    def buildTree2(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        def build(i1, i2, p1, p2):
            if i1 > i2:
                return None
            root = TreeNode(postorder[p2])
            root_idx = inorder.index(postorder[p2])
            root.left = build(i1, root_idx - 1, p1, p1 + root_idx - i1 - 1)
            root.right = build(root_idx + 1, i2, p1 + root_idx - i1, p2 - 1)
            return root
        
        return build(0, len(inorder) - 1, 0, len(postorder) - 1)
        